File: FREQNESS_Toolbox/FREQNESS_Functions/FREQNESS_ComputeFilterWidths.m
function fwidth = FREQNESS_ComputeFilterWidths(frex,filter)

% ========================================================================
%
%  FREQUENCY-RESOLVED NETWORK ESTIMATION TOOLBOX
%
%  If you use this toolbox, please cite:
%  Rosso, M., Fernández‐Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
%  Kringelbach, M. L., & Bonetti, L. (2025).
%  FREQ‐NESS Reveals the Dynamic Reconfiguration of Frequency‐Resolved Brain
%  Networks During Auditory Stimulation. Advanced Science, 2413195.
%  https://doi.org/10.1002/advs.202413195
%
% ========================================================================
%  This function computes physically calibrated Gaussian-filter widths
%  from the analyzed frequencies. The schedule depends on frequency in Hz
%  and its numerical range, not on the number or spacing of frequency bins.
%
%  The logarithmic schedule (default) decreases spectral selectivity
%  smoothly from Q = 7 at the lowest frequency to Q = 3.5 at the highest:
%
%      t(f) = log(f/fmin) / log(fmax/fmin)
%      Q(f) = 7 * (3.5/7)^t(f)
%      FWHM(f) = f / Q(f)
%
%  The linear schedule uses constant Q = 5, so FWHM grows linearly with
%  centre frequency. For a single unique frequency, the logarithmic mode
%  uses the geometric midpoint Q = sqrt(7*3.5).
%
% ------------------------------------------------------------------------
%  INPUTS:
% ------------------------------------------------------------------------
%  - frex   : Non-empty vector of positive finite frequencies in Hz.
%  - filter : 'logarithmic' (default) or 'linear'.
%
% ------------------------------------------------------------------------
%  OUTPUT:
% ------------------------------------------------------------------------
%  - fwidth : Row vector containing one positive FWHM value in Hz for each
%             input frequency, in the same element order as frex.
%
% ------------------------------------------------------------------------
%  AUTHORS:
%  Mattia Rosso & Leonardo Bonetti
%  mattia.rosso@clin.au.dk
%  leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
%
% ========================================================================

if nargin < 1 || ~isnumeric(frex) || ~isreal(frex) || ...
        ~isvector(frex) || isempty(frex) || any(~isfinite(frex)) || ...
        any(frex <= 0)
    error('frex must be a non-empty vector of positive finite frequencies.');
end

if nargin < 2 || isempty(filter)
    filter = 'logarithmic';
end

if ~(ischar(filter) || (isstring(filter) && isscalar(filter))) || ...
        ~(strcmpi(filter,'logarithmic') || strcmpi(filter,'linear'))
    error('filter must be either ''logarithmic'' or ''linear''.');
end

frex = double(frex(:)');

if strcmpi(filter,'linear')
    q_values = 5 * ones(size(frex));
else
    q_low = 7;
    q_high = 3.5;
    fmin = min(frex);
    fmax = max(frex);

    if fmin == fmax
        q_values = sqrt(q_low*q_high) * ones(size(frex));
    else
        log_position = log(frex./fmin) ./ log(fmax/fmin);
        q_values = q_low .* (q_high/q_low).^log_position;
    end
end

fwidth = frex ./ q_values;

end

File: FREQNESS_Toolbox/python/src/freqness/FREQNESS_ComputeFilterWidths.py
"""Physically calibrated Gaussian-filter widths for FREQ-NESS.

If you use this toolbox, please cite:
Rosso, M., Fernández-Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
Kringelbach, M. L., & Bonetti, L. (2025). FREQ-NESS Reveals the Dynamic
Reconfiguration of Frequency-Resolved Brain Networks During Auditory
Stimulation. Advanced Science, 2413195.
https://doi.org/10.1002/advs.202413195

Toolbox authors
---------------
Mattia Rosso - Center for Music in the Brain, Aarhus University -
mattia.rosso@clin.au.dk
Chiara Malvaso - University of Bologna - chiara.malvaso2@unibo.it
Leonardo Bonetti - Center for Music in the Brain, Aarhus University; Centre
for Eudaimonia and Human Flourishing, Linacre College, University of Oxford -
leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
"""

from __future__ import annotations

import numpy as np
from numpy.typing import ArrayLike, NDArray


FloatArray = NDArray[np.float64]


def FREQNESS_ComputeFilterWidths(
    frex: ArrayLike,
    filter: str = "logarithmic",
) -> FloatArray:
    """Compute automatic FWHM values from physical frequencies in Hz.

    The default logarithmic schedule varies selectivity from Q = 7 at the
    lowest requested frequency to Q = 3.5 at the highest. The linear schedule
    uses constant Q = 5. A single unique frequency in logarithmic mode uses
    the geometric midpoint Q = sqrt(7 * 3.5). Output order matches ``frex``.
    """
    raw_frequencies = np.asarray(frex)
    if raw_frequencies.ndim == 0:
        raw_frequencies = raw_frequencies.reshape(1)
    if (
        raw_frequencies.ndim != 1
        or raw_frequencies.size == 0
        or not np.issubdtype(raw_frequencies.dtype, np.number)
        or np.iscomplexobj(raw_frequencies)
    ):
        raise ValueError(
            "frex must be a non-empty vector of positive finite frequencies"
        )

    frequencies = np.asarray(raw_frequencies, dtype=float)
    if not np.all(np.isfinite(frequencies)) or np.any(frequencies <= 0):
        raise ValueError(
            "frex must be a non-empty vector of positive finite frequencies"
        )
    if not isinstance(filter, str) or filter.lower() not in {
        "logarithmic",
        "linear",
    }:
        raise ValueError("filter must be either 'logarithmic' or 'linear'")

    if filter.lower() == "linear":
        q_values = np.full(frequencies.shape, 5.0)
    else:
        q_low = 7.0
        q_high = 3.5
        minimum = float(np.min(frequencies))
        maximum = float(np.max(frequencies))

        if minimum == maximum:
            q_values = np.full(
                frequencies.shape,
                np.sqrt(q_low * q_high),
            )
        else:
            log_position = np.log(frequencies / minimum) / np.log(
                maximum / minimum
            )
            q_values = q_low * (q_high / q_low) ** log_position

    return frequencies / q_values


__all__ = ["FREQNESS_ComputeFilterWidths"]
